File: interview_cake/dynamic_programming_and_recursion/permutations.rb
require 'set'

def get_permutations(string)

  # Generate all permutations of the input string.
  return Set.new([string]) if string.length <= 1
  
  permutations = Set.new
  sub_permutations = get_permutations(string[1..-1])
  sub_permutations.each do |sub_permutation|
    0.upto(sub_permutation.length) do |n|
      permutations.add(sub_permutation.dup.insert(n, string[0]))
    end
  end
  permutations
end


















# Tests

def run_tests
  desc = 'empty string'
  actual = get_permutations('')
  expected = Set.new([''])
  assert_equal(actual, expected, desc)

  desc = 'one character string'
  actual = get_permutations('a')
  expected = Set.new(['a'])
  assert_equal(actual, expected, desc)

  desc = 'two character string'
  actual = get_permutations('ab')
  expected = Set.new(['ab', 'ba'])
  assert_equal(actual, expected, desc)

  desc = 'three character string'
  actual = get_permutations('abc')
  expected = Set.new(['abc', 'acb', 'bac', 'bca', 'cab', 'cba'])
  assert_equal(actual, expected, desc)
end

def assert_equal(a, b, desc)
  puts "#{desc} ... #{a == b ? 'PASS' : "FAIL: #{a.inspect} != #{b.inspect}"}"
end

run_tests

File: interview_cake/dynamic_programming_and_recursion/permutations.js
function getPermutations(string) {

  // Generate all permutations of the input string
  if (string.length <= 1) {
    return new Set([string]);
  }
  
  let permutations = new Set();
  let subPermutations = getPermutations(string.slice(1, string.length));
  subPermutations.forEach(subPermutation => {
    for (let position = 0; position <= subPermutation.length; position++) {
      permutation = subPermutation.slice(0, position) + string[0] + 
                    subPermutation.slice(position, subPermutation.length);
      permutations.add(permutation);
    }
  });
  
  return permutations;
}


















// Tests

let desc = 'empty string';
let input = '';
let actual = getPermutations(input);
let expected = new Set(['']);
assert(isSetsEqual(actual, expected), desc);

desc = 'one character string';
input = 'a';
actual = getPermutations(input);
expected = new Set(['a']);
assert(isSetsEqual(actual, expected), desc);

desc = 'two character string';
input = 'ab';
actual = getPermutations(input);
expected = new Set(['ab', 'ba']);
assert(isSetsEqual(actual, expected), desc);

desc = 'three character string';
input = 'abc';
actual = getPermutations(input);
expected = new Set(['abc', 'acb', 'bac', 'bca', 'cab', 'cba']);
assert(isSetsEqual(actual, expected), desc);

function isSetsEqual(as, bs) {
  if (as.size !== bs.size) {
    return false;
  }
  for (let a of as) {
    if (!bs.has(a)) return false;
  }
  return true;
}

function assert(condition, desc) {
  if (condition) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL`);
  }
}
